File: network.py
import math
import random
import matplotlib.pyplot as plt

class Activation:
  class ReLu:
    def compute(self, z):
      return [z[i] if z[i] > 0 else 0 for i in range(len(z))]
  
    def derivative(self, z):
      return [1 if z[i] > 0 else 0 for i in range(len(z))]

  class LeakyReLu:
    def __init__(self, eta):
      self.eta = eta

    def compute(self, z):
      return [z[i] if z[i] > 0 else z[i] * self.eta for i in range(len(z))]
  
    def derivative(self, z):
      return [1 if z[i] > 0 else self.eta for i in range(len(z))]

  class TanH:
    def compute(self, z):
      a = [0 for i in range(len(z))]
      for i in range(len(z)):
        e2 = math.exp(2 * z[i])
        a[i] = (e2 - 1) / (e2 + 1)
      return a

    def derivative(self, z):
      t = self.compute(z)
      return [1 - t[i] * t[i] for i in range(len(z))]

class Loss:
  class MeanSquaredError:
    def compute(self, x, y):
      loss = 0
      for i in range(len(x)):
        error = x[i] - y[i]
        loss += error ** 2
      return loss * 0.5

    def derivative(self, x, y):
      return [x[i] - y[i] for i in range(len(x))]

class Network:
  def __init__(self, layerSizes, activation):
    self.layerSizes = layerSizes
    self.activation = activation
    self.numLayers = len(self.layerSizes)
    self.weights = [[[random.normalvariate(0, 1) for k in range(self.layerSizes[l])] for j in range(self.layerSizes[l - 1])] for l in range(1, self.numLayers)]
    self.biases = [[0 for j in range(self.layerSizes[l])] for l in range(0, self.numLayers)]

  def feedForward(self, x):
    for l in range(1, self.numLayers):
      z = [0 for i in range(self.layerSizes[l])]
      for j in range(self.layerSizes[l]):
        z[j] = self.biases[l][j]
        for k in range(self.layerSizes[l - 1]):
          z[j] += x[k] * self.weights[l - 1][k][j]
      x = self.activation.compute(z)
    return x

  def backPropagate(self, x, y, loss):
    zs = [None]
    activations = [x]

    for l in range(1, self.numLayers):
      z = [0 for i in range(self.layerSizes[l])]
      for j in range(self.layerSizes[l]):
        z[j] = self.biases[l][j]
        for k in range(self.layerSizes[l - 1]):
          z[j] += x[k] * self.weights[l - 1][k][j]
      x = self.activation.compute(z)

      zs.append(z)
      activations.append(x)

    delta = [[0 for j in range(self.layerSizes[l])] for l in range(0, self.numLayers)]

    dLoss = loss.derivative(activations[self.numLayers - 1], y)
    sp = self.activation.derivative(zs[self.numLayers - 1])
    delta[self.numLayers - 1] = [dLoss[i] * sp[i] for i in range(self.layerSizes[self.numLayers - 1])]

    for l in range(self.numLayers - 2, 0, -1):
      sum = [0 for i in range(self.layerSizes[l])]
      for j in range(self.layerSizes[l]):
        for k in range(self.layerSizes[l + 1]):
          sum[j] += self.weights[l][j][k] * delta[l + 1][k]
      sp = self.activation.derivative(zs[l])
      delta[l] = [sum[i] * sp[i] for i in range(self.layerSizes[l])]

    nablaW = [[[0 for k in range(self.layerSizes[l])] for j in range(self.layerSizes[l - 1])] for l in range(1, self.numLayers)]
    nablaB = [[0 for j in range(self.layerSizes[l])] for l in range(0, self.numLayers)]

    for l in range(0, self.numLayers - 1):
      for j in range(self.layerSizes[l]):
        for k in range(self.layerSizes[l + 1]):
          nablaW[l][j][k] = activations[l][j] * delta[l + 1][k]
    for l in range(0, self.numLayers):
      for j in range(self.layerSizes[l]):
        nablaB[l][j] = delta[l][j]
    
    return (nablaW, nablaB)

  def train(self, loss, trainingData, epochs, eta):
    for i in range(epochs):
      for j in range(len(trainingData)):
        nablaW, nablaB = self.backPropagate(trainingData[j][0], trainingData[j][1], loss)

        for l in range(0, self.numLayers - 1):
          for j in range(self.layerSizes[l]):
            for k in range(self.layerSizes[l + 1]):
              self.weights[l][j][k] -= nablaW[l][j][k] * eta
        for l in range(0, self.numLayers):
          for j in range(self.layerSizes[l]):
            pass
            self.biases[l][j] -= nablaB[l][j] * eta



def shuffle(arr):
  n = len(arr)
  for i in range(n-1,0,-1):
    j = random.randint(0,i+1)
    arr[i],arr[j] = arr[j],arr[i]
  return arr

trainingData = [[[math.cos(i / 20 * math.tau) * 0.15 + (random.normalvariate(0, 0.1)), math.sin(i / 20 * math.tau) * 0.15 + (random.normalvariate(0, 0.1))], [-1]] for i in range(20)] + [[[math.cos(i/50 * math.tau) * 0.45 + (random.normalvariate(0, 0.1)), math.sin(i/50 * math.tau) * 0.45 + (random.normalvariate(0, 0.1))], [1]] for i in range(50)] + [[[math.cos(i / 80 * math.tau) * 0.75 + (random.normalvariate(0, 0.1)), math.sin(i / 80 * math.tau) * 0.75 + (random.normalvariate(0, 0.1))], [-1]] for i in range(80)]
trainingData = shuffle(trainingData)

network = Network((2, 8, 8, 1), Activation.TanH())
loss = Loss.MeanSquaredError()

network.train(loss, trainingData, 100, 0.05)

x0 = []
y0 = []
x1 = []
y1 = []

for i in range(80):
  for j in range(80):
    x = i / 40 - 1
    y = j / 40 - 1
    output = network.feedForward([x, y])[0]
    if output < 0:
      x0.append(x)
      y0.append(y)
    else:
      x1.append(x)
      y1.append(y)

x2 = []
y2 = []
x3 = []
y3 = []

for i in range(len(trainingData)):
  if (trainingData[i][1][0] < 0):
    x2.append(trainingData[i][0][0])
    y2.append(trainingData[i][0][1])
  else:
    x3.append(trainingData[i][0][0])
    y3.append(trainingData[i][0][1])

fig, ax = plt.subplots()
ax.plot(x0, y0, 'o')
ax.plot(x1, y1, 'o')
ax.plot(x2, y2, 'o')
ax.plot(x3, y3, 'o')
ax.set(xlim=(-1, 1), ylim=(-1, 1))
plt.show()

File: network.js
class Utils {
  static gaussianRandom(mean, stddev) {
    // https://stackoverflow.com/questions/25582882/javascript-math-random-normal-distribution-gaussian-bell-curve
    return Math.sqrt(-2 * Math.log(1 - Math.random())) * Math.cos(2 * Math.PI * Math.random()) * stddev + mean;
  }
}

class Activation {
  static TanH = class {
    compute(z) {
      let a = new Float64Array(z.length);
      for (let j = 0; j < z.length; j++) {
        let e2 = Math.exp(2 * z[j]);
        a[j] = (e2 - 1) / (e2 + 1);
      }
      return a;
    }

    derivative(z) {
      let t = this.compute(z);
      let d = new Float64Array(z.length);
      for (let j = 0; j < z.length; j++)
        d[j] = 1 - t[j] * t[j];
      return d;
    }
  }

  static LeakyReLu = class {
    constructor(alpha) {
      this.alpha = alpha;
    }

    compute(z) {
      let a = new Float64Array(z.length);
      for (let j = 0; j < z.length; j++)
        a[j] = z[j] > 0 ? z[j] : z[j] * this.alpha;
      return a;
    }

    derivative(z) {
      let d = new Float64Array(z.length);
      for (let j = 0; j < z.length; j++)
        d[j] = z[j] > 0 ? 1 : this.alpha;
      return d;
    }
  }
}

class Cost {
  static MeanSquaredError = class {
    compute(a, y) {
      let cost = 0;
      for (let j = 0; j < a.length; j++) {
        error = a[j] - y[j];
        cost += error * error;
      }
      return cost * 0.5;
    }

    derivative(a, y) {
      let d = new Float64Array(a.length);
      for (let j = 0; j < a.length; j++)
        d[j] = a[j] - y[j];
      return d;
    }
  }
}

class Network {
  constructor(layerSizes, activation, cost) {
    this.layerSizes = layerSizes;
    this.activation = activation;
    this.cost = cost;

    this.numLayers = this.layerSizes.length;
    this.weights = new Array();
    for (let l = 1; l < this.numLayers; l++) {
      let array0 = new Array();
      for (let k = 0; k < this.layerSizes[l - 1]; k++) {
        let array1 = new Float64Array(this.layerSizes[l]);
        for (let j = 0; j < this.layerSizes[l]; j++)
          array1[j] = Utils.gaussianRandom(0, 1) / Math.sqrt(this.layerSizes[l - 1]);
        array0.push(array1);
      }
      this.weights.push(array0);
    }
    this.biases = new Array();
    for (let l = 0; l < this.numLayers; l++) {
      let array0 = new Float64Array(this.layerSizes[l]);
      this.biases.push(array0);
    }

    this.gradientW = new Array();
    for (let l = 1; l < this.numLayers; l++) {
      let array0 = new Array();
      for (let k = 0; k < this.layerSizes[l - 1]; k++) {
        let array1 = new Float64Array(this.layerSizes[l]);
        array0.push(array1);
      }
      this.gradientW.push(array0);
    }
    this.gradientB = new Array();
    for (let l = 0; l < this.numLayers; l++) {
      let array0 = new Float64Array(this.layerSizes[l]);
      this.gradientB.push(array0);
    }
  }

  feedForward(a) {
    for (let l = 1; l < this.numLayers; l++) {
      let z = new Float64Array(this.layerSizes[l]);
      for (let j = 0; j < this.layerSizes[l]; j++) {
        z[j] = this.biases[l][j];
        for (let k = 0; k < this.layerSizes[l - 1]; k++)
          z[j] += a[k] * this.weights[l - 1][k][j];
      }
      a = this.activation.compute(z);
    }
    return a;
  }

  backPropagate(a, y) {
    let zs = [null];
    let as = [a];
    for (let l = 1; l < this.numLayers; l++) {
      let z = new Float64Array(this.layerSizes[l]);
      for (let j = 0; j < this.layerSizes[l]; j++) {
        z[j] = this.biases[l][j];
        for (let k = 0; k < this.layerSizes[l - 1]; k++)
          z[j] += a[k] * this.weights[l - 1][k][j];
      }
      a = this.activation.compute(z);
      zs.push(z);
      as.push(a);
    }
    let error = new Array();
    for (let l = 0; l < this.numLayers; l++) {
      let array0 = new Float64Array(this.layerSizes[l]);
      for (let j = 0; j < this.layerSizes[l]; j++)
        array0[j] = 0;
      error.push(array0);
    }
    let aPrime = this.activation.derivative(zs[this.numLayers - 1]);
    let cPrime = this.cost.derivative(as[this.numLayers - 1], y);
    for (let j = 0; j < this.layerSizes[this.numLayers - 1]; j++)
      error[this.numLayers - 1][j] = aPrime[j] * cPrime[j];
    for (let l = this.numLayers - 1; l > 1; l--) {
      let sum = new Float64Array(this.layerSizes[l - 1]);
      for (let k = 0; k < this.layerSizes[l - 1]; k++)
        for (let j = 0; j < this.layerSizes[l]; j++)
          sum[k] += this.weights[l - 1][k][j] * error[l][j];
      let aPrime = this.activation.derivative(zs[l - 1]);
      for (let k = 0; k < this.layerSizes[this.numLayers - 1]; k++)
        error[l - 1][k] = sum[k] * aPrime[k];
    }
    for (let l = 1; l < this.numLayers; l++)
      for (let k = 0; k < this.layerSizes[l - 1]; k++)
        for (let j = 0; j < this.layerSizes[l]; j++)
          this.gradientW[l - 1][k][j] = as[l - 1][k] * error[l][j];
    for (let l = 0; l < this.numLayers; l++)
      for (let j = 0; j < this.layerSizes[l]; j++)
        this.gradientB[l][j] = error[l][j];
  }

  applyGradients(eta) {
    for (let l = 1; l < this.numLayers; l++)
      for (let k = 0; k < this.layerSizes[l - 1]; k++)
        for (let j = 0; j < this.layerSizes[l]; j++)
          this.weights[l - 1][k][j] -= this.gradientW[l - 1][k][j] * eta;
    for (let l = 0; l < this.numLayers; l++)
      for (let j = 0; j < this.layerSizes[l]; j++)
        this.biases[l][j] -= this.gradientB[l][j] * eta
  }
}



let network = new Network(
  [1, 1],
  new Activation.TanH(),
  new Cost.MeanSquaredError());